### PR TITLE
Extend BUILD_PATH_PREFIX_MAP with the workspace root

### PR DESCRIPTION
**Pull request by jeremiedimino**
For better reproducibility

### CODE REVIEW

**Comment by jeremiedimino** [comment]
This is still a WIP, I need to fix the test

**Comment by jeremiedimino** [comment]
Alright, this is ready for review.

File: src/dune_engine/action_exec.mli [hunk 1]
```diff
@@ -27,12 +27,33 @@ module Exec_result : sig
   type t = { dynamic_deps_stages : (Dynamic_dep.Set.t * Dep.Facts.t) List.t }
 end
 
+(** [root] shoubd be the root of the current build context, or the root of the
```
Comment (snowleopard) [inline src/dune_engine/action_exec.mli:30]:
```suggestion
(** [root] should be the root of the current build context, or the root of the
```

File: src/dune_engine/action_exec.ml [hunk 2]
```diff
@@ -516,8 +516,29 @@ let exec_until_all_deps_ready ~ectx ~eenv t =
...
+let _BUILD_PATH_PREFIX_MAP = "BUILD_PATH_PREFIX_MAP"
+
+let extend_build_path_prefix_map env how map =
+  let new_rules = Build_path_prefix_map.encode_map map in
+  Env.update env ~var:_BUILD_PATH_PREFIX_MAP ~f:(function
+    | None -> Some new_rules
+    | Some existing_rules ->
+      Some
+        (match how with
+        | `Existing_rules_have_precedence -> new_rules ^ ":" ^ existing_rules
```
Comment (snowleopard) [inline src/dune_engine/action_exec.ml:528]: Looks like all call sites use `New_rules_have_precedence`, so this code path is currently dead.

**Review by snowleopard — APPROVED** [review]
Looks good!

File: src/dune_engine/action_exec.ml (hunk 2 above)
Comment (jeremiedimino) [reply src/dune_engine/action_exec.ml:528]: Indeed. I think having the argument is useful to force callers to think about it. We could remove the `Existing_rules_have_precedence` case, though the API feels somewhat incomplete without it, and support is just a one liner.

File: src/dune_engine/action_exec.mli [hunk 3]
```diff
@@ -27,12 +27,33 @@ module Exec_result : sig
...
+
+    Note that the rewriting rules are applied from right to left, so the last
+    rule of [map] will be tried first.
+
+    If the environment variable is already defined in [env], [how] explains
+    whether the rules in [map] should be tried before or after the existing
+    ones. *)
+val extend_build_path_prefix_map :
+     Env.t
+  -> [ `Existing_rules_have_precedence | `New_rules_have_precedence ]
```
Comment (rgrinberg) [inline src/dune_engine/action_exec.mli:57]: Polymorphic variants are back "in". Nice :)

**Comment by rgrinberg** [comment]
@jeremiedimino i only had a cursory glance at this, but shouldn't this change be versioned?

**Comment by rgrinberg** [comment]
Oh and a CHANGES entry.

File: src/dune_engine/action_exec.mli (hunk 3 above)
Comment (jeremiedimino) [reply src/dune_engine/action_exec.mli:57]: Yeah, I gave in

**Comment by jeremiedimino** [comment]
> @jeremiedimino i only had a cursory glance at this, but shouldn't this change be versioned?

I wondered about it, but after all this only affects command that already know how to deal with this. I don't feel strongly about it though, happy to version it.

**Comment by rgrinberg** [comment]
Then let’s skip it. No need to complicate the code without a good reason.

Rudi.

**Comment by jeremiedimino** [comment]
> Oh and a CHANGES entry.

Pushed to main

File: src/dune_engine/action_exec.mli (hunk 3 above)
Comment (snowleopard) [reply src/dune_engine/action_exec.mli:57]: Yeah, I followed the lead and sneaked in a polymorphic variant too :)

https://github.com/ocaml/dune/blob/65404cf973aa7ffc0e9dd7d05c9dd3709c7db2d4/otherlibs/stdune-unstable/temp.ml#L14-L20

**Pull request merged**